src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -39,7 +39,8 @@ public static IList<Diagnostic> GetCompileTimeConstantViolation(SyntaxBase expre
         /// <returns>Returns true if values of the specified source type are assignable to the target type. Returns false otherwise or null if assignability cannot be determined.</returns>
         public static bool? AreTypesAssignable(TypeSymbol? sourceType, TypeSymbol? targetType)
         {
-            if (sourceType == null || targetType == null || sourceType is UnassignableTypeSymbol || targetType is UnassignableTypeSymbol)
+            if (sourceType == null || sourceType.ValidationFlags.HasFlag(TypeSymbolValidationFlags.PreventAssignment) ||
+                targetType == null || targetType.ValidationFlags.HasFlag(TypeSymbolValidationFlags.PreventAssignment))

```

Comment:
The return values are supposed to work like this:
true - definitely assignable
false - definitely not assignable
null - who knows? 🤷‍♀️ (only really possible due to missing info)

If the PreventAssignment flag is present, I think we should return false.

Reply:
Good point

Reply:
I've removed null as an option here